File: 34/main.py
########################################################################################
############	34. Find First and Last Position of Element in Sorted Array	########
########################################################################################



'''
Perform binary search on first half and second half of array. Return the maximum index from calls
Time Complexity:O(logn) because we do binary search
Space Complexity:O(1) because we store the indices
'''


class Solution:
    def searchRange(self, nums, target):
        if not nums:
            return [-1,-1]
        def search(lo, hi):
            if nums[lo] == target == nums[hi]:
                return [lo, hi]
            if nums[lo] <= target <= nums[hi]:
                mid = (lo + hi) // 2
                l, r = search(lo, mid), search(mid+1, hi)
                return max(l, r) if -1 in l+r else [l[0], r[1]]
            return [-1, -1]
        return search(0, len(nums)-1)

File: 34/main.java
/**
*
*	34. Find First and Last Position of Element in Sorted Array
*
**/



/*
This solution is mirrored version of python solution. Refer to python solution for approach
*/

public class Solution {
	public int[] searchRange(int[] nums, int target) {
		int start = Solution.firstGreaterEqual(nums, target);
		if (start == nums.length || nums[start] != target) {
			return new int[]{-1, -1};
		}
		return new int[]{start, Solution.firstGreaterEqual(nums, target + 1) - 1};
	}

	private static int firstGreaterEqual(int[] A, int target) {
		int low = 0, high = A.length;
		while (low < high) {
			int mid = low + ((high - low) >> 1);
			
			if (A[mid] < target) {
				low = mid + 1;
			} else {
				high = mid;
			}
		}
		return low;
	}
}     

